### ISSUE TITLE
Start measuring Tekton Pipelines performance

### ISSUE THREAD

**Issue by bobcatfish** [knative/build-pipeline#540]
## Expected Behavior

We should be measuring performance for Pipelines. This task includes both adding the actual measurement mechanism and also the design re. what exactly we want to measurement.

Some ideas for measurement:

- Null Task / null Pipeline (i.e. it doesnt actually do anything)
- Null Tasks that have linked inputs outputs
- ?

## Actual Behavior

We do not measure or track this.

## Additional Info

* Other knative projects use this https://testgrid-dot-knative-tests.appspot.com/knative-build#latency
* Note that the metrics collector being used by these projects is global and our end to end tests run in parallel, so the metrics will not work as is

**Comment by pradeepitm12** [comment]
Hello @bobcatfish 
Need your thoughts on this. 
1-  A service outside of tekton that watches tekton object and expose it to prometheus.
2- Introduce an endpoint in the tekton pipeline itself to expose all the metric to Prometheus.

**Comment by bobcatfish** [comment]
My gut feeling is that I'd lean more toward exposing the metrics from Pipelines itself:

> 2- Introduce an endpoint in the tekton pipeline itself to expose all the metric to Prometheus.

Question: I'm not super familair with Prometheus, how vital would it be to making metrics usable? Could we simply emit the metrics, and allow the user to provide their own metrics gathering mechanism (which could be prometheus but could be something else), or would it make more sense for us to include Prometheus out of the box? (I've very sensitive to adding new dependencies, esp. since I'm under the impression that managing Prometheus is a job in itself, but maybe I'm wrong!)

Another option, which I think is a variation on your first suggestion @pradeepitm12 :
3 - (For now) only measure the performance in tests we write specifically for this purpose (i.e. we don't expose anything new for users of Tekton Pipelines, but we start doing our own measurements)

**Comment by rawlingsj** [comment]
+1 we're looking at the same thing and just started looking at prometheus too, hopefully we can help each other out here.

**Comment by bobcatfish** [comment]
> +1 we're looking at the same thing and just started looking at prometheus too, hopefully we can help each other out here.

Maybe the first thing to do would be to identify the metrics we're interested in? I'm not super familiar with prometheus but I would think before we want to monitor the metrics, we'd want to figure out what needs monitoring (maybe there's a Jenkins/Jenkins X precedent we can draw on :D?)

**Comment by sbwsg** [comment]
We had our first meeting regarding observability, specifically metrics, today and work is now underway. There are a couple of other issues that overlap in theme with this one. I am linking them together here for us to review later and figure out which to keep and which to close.

Related issues:
https://github.com/tektoncd/pipeline/issues/164
https://github.com/tektoncd/pipeline/issues/540
https://github.com/tektoncd/pipeline/issues/855

[Metrics Design Doc](https://docs.google.com/document/d/1NbY3spcNVXqrZHbEp5pfFQbCyGtN8Aj7Y5d4elwM5cI/edit)

[Notes from the initial metrics meeting](https://docs.google.com/document/d/1n-ZJzKMEH31mDSooaCjVV5b3d19zwJkuDm4fmrevoLs/edit)

**Comment by bobcatfish** [comment]
We haven't worked on this lately but it is an item in [our roadmap](https://github.com/tektoncd/pipeline/issues/540) and I think we should keep it open.

/lifecycle frozen

**Comment by bobcatfish** [comment]
I want to start gathering some requirements around this and get it moving :D

**Comment by bobcatfish** [comment]
https://github.com/tektoncd/pipeline/issues/3521 has some use cases that we might be able use

**Comment by mengjieli0726** [comment]
@bobcatfish, any tektone performance white paper have?  as so far,  how many pipeline run or run we can support in middle cluster (just like: 1 master + 1 compute node.)
the node spec:  8 core + 64 G memory + 250 G disk.

**Comment by beanscg** [comment]
Bean Labs took a public-source pass on this issue.

The narrowest useful next step looks like: Offer to build a minimal repro or fixture around `cloud.go` and report the exact observed behavior.

Public evidence checked:
- Open public issue in `tektoncd/pipeline`; repo license `Apache-2.0`; no assignee; 10 comments.
- Labels: kind/feature, design, meaty-juicy-coding-work, okr, lifecycle/frozen, priority/important-longterm.
- Referenced files/paths: cloud.go.

A good acceptance check would be a narrow public change that directly exercises the issue's reported boundary.

No PR opened from this pass.

**Comment by vdemeester** [comment]
Closing in favor of #9152 (TEP-0036: Start Measuring Tekton Pipelines Performance) which is the formal TEP tracking the same goal.
